### PR TITLE
[RFC] Fix msg_puts_printf() on windows

### PR DESCRIPTION
**Pull request by erw7**
This PR fixes the following issues:

- `nvim --version`: NVIM v0.4.0-657-ga0d723db5
- Vim (version: ) behaves differently? Yes(8.1.1397)
- Operating system/version: Microsoft Windows [Version 6.1.7601]
- Terminal name/version:
- `$TERM`:

### Steps to reproduce using `nvim -u NORC`

```
set LANG=ja_JP
nvim -u NORC --cmd "echomsg 'Hello'"
```

### Actual behaviour
Messages are displayed in garbled.

```
Hello
邯壹￠繧九↓縺ｯENTER繧呈款縺吶°繧ｳ繝槭Φ繝峨ｒ蜈･蜉帙＠縺ｦ縺上□縺輔＞
```

### Expected behaviour
The message is displayed normally.

```
Hello
続けるにはENTERを押すかコマンドを入力してください
```

### CODE REVIEW

File: src/nvim/vim.h [hunk 1]
```diff
@@ -287,9 +291,34 @@ enum { FOLD_TEXT_LEN = 51 };  //!< buffer size for get_foldtext()
 // functions of these names. The declarations would break if the defines had
 // been seen at that stage.  But it must be before globals.h, where error_ga
 // is declared.
-#define mch_errmsg(str)        fprintf(stderr, "%s", (str))
+#ifdef WIN32
+# define mch_errmsg(str) \
```
Comment (bfredl) [inline src/nvim/vim.h]: Why not write these as functions?

**Comment by justinmk** [comment]
Is this needed after https://github.com/neovim/neovim/pull/9829? Would much rather remove `USE_MCH_ERRMSG` entirely and go forward with https://github.com/neovim/neovim/pull/9829. If it's blocked by https://github.com/neovim/neovim/pull/9338 I will try to unblock it .

File: test/functional/core/startup_spec.lua [hunk 2]
```diff
@@ -219,6 +219,29 @@ describe('startup', function()
     clear{args={'--embed'}}
     eq(2, eval('1+1'))
   end)
+
+  it('message display ends with enter #7967', function()
```
Comment (justinmk) [inline test/functional/core/startup_spec.lua]: #7967 is fixed on master, right? So we can add this test already.

**Comment by erw7** [comment]
> Is this needed after #9829?

Even after #9829 `nvim -V1 -Es` needs this. For #9338 I created #10109.

File: test/functional/core/startup_spec.lua (hunk 2 above)
Comment (erw7) [reply test/functional/core/startup_spec.lua]: Yes, #7967 has been fixed by # 9884.

**Comment by justinmk** [comment]
> Even after #9829 `nvim -V1 -Es` needs this. 

If it's not much trouble, can you add a test for that?

File: src/nvim/vim.h [hunk 3]
```diff
@@ -287,9 +287,11 @@ enum { FOLD_TEXT_LEN = 51 };  //!< buffer size for get_foldtext()
 // functions of these names. The declarations would break if the defines had
 // been seen at that stage.  But it must be before globals.h, where error_ga
 // is declared.
-#define mch_errmsg(str)        fprintf(stderr, "%s", (str))
+#ifndef WIN32
+# define mch_errmsg(str)        fprintf(stderr, "%s", (str))
+# define mch_msg(str)           printf("%s", (str))
+#endif
 #define display_errors()       fflush(stderr)
```
Comment (justinmk) [inline src/nvim/vim.h]: `display_errors` could be removed completely. It is only used in `do_shell()`, left over from Vim. `do_shell()` in Nvim uses the Nvim UI, not the tty directly.

**Comment by erw7** [comment]
>> Even after #9829 nvim -V1 -Es needs this.

> If it's not much trouble, can you add a test for that?

@justinmk  The test environment does not use message catalogs, because `LOCALE_INSTALL_DIR` is passed to` bindtextdomain()`. This is also true for appimage and Windows  artifacts. We need to solve the problem before creating a test.

I'm thinking of changing to `exepath/../ share/locale` instead of `LOCALE_INSTALL_DIR`, like `os_setenv_append_path()`. Do you have any other good ideas? Do I need to create a new issue on this matter?

**Comment by justinmk** [comment]
> I'm thinking of changing to `exepath/../ share/locale` instead of `LOCALE_INSTALL_DIR`, like `os_setenv_append_path()`. 

:+1: that seems completely fine.

**Comment by justinmk** [comment]
```
../src/nvim/main.c:724:61: error: implicit conversion changes signedness: 'long' to 'unsigned long' [-Werror,-Wsign-conversion]
  xstrlcpy(tail, "share/locale", sizeof(localepath) - (tail - localepath));
                                                    ~  ~~~~~^~~~~~~~~~~~
```

File: src/nvim/main.c [hunk 4]
```diff
@@ -738,22 +738,17 @@ static void init_locale(void)
...
+  size_t exepathlen = MAXPATHL;
+  if (os_exepath(exepath, &exepathlen) != 0) {
+    path_guess_exepath(argv0 ? argv0 : "nvim", exepath, sizeof(exepath));
   }
-# endif
+  char *tail = (char *)path_tail_with_sep((char_u *)exepath);
+  *tail = NUL;
+  tail = (char *)path_tail((char_u *)exepath);
+  xstrlcpy(tail, "share/locale", sizeof(localepath) - (tail - localepath));
+  bindtextdomain(PROJECT_NAME, localepath);
```
Comment (justinmk) [inline src/nvim/main.c]: does not regress 1a15cf84c29c984d600f438122edc70be2ac91a1 as long as we assume `@CMAKE_INSTALL_FULL_LOCALEDIR@` would always be in this form.

I wonder if this conforms with BSD conventions.

**Pull request merged**